File: lectures/01-21/fact.ml
let rec fact (n : int) : int =
  if n = 0
  then 1
  else n * fact (n - 1)

let _ = assert (fact 5 = 120)

let fact n =
  let acc = 1 in
  let i = 1 in
  let rec loop acc i =
    let acc = acc * i in
    let i = i + 1 in
    if i <= n
    then loop acc i (* no return statement *)
    else acc
  in loop acc i

let _ = assert (fact 5 = 120)

let fact n =
  let rec loop acc i =
    if i <= n
    then loop (acc * i) (i + 1)
    else acc
  in loop 1 1

(*

fact 5
loop 1 1
if 1 <= 5 then loop (1 * 1) (1 + 1) else 1
loop 1 2
loop 2 3
loop 6 4
loop 24 5
loop 120 6
if 6 <= 5 then loop (120 * 6) (6 + 1) else 120
120

*)

File: lectures/01-21/fact.py
def fact(n):
  acc = 1
  for i in range(1, n + 1):
    acc = acc * i
  return acc

assert(fact(5) == 120)

def fact(n):
  if n <= 0:
    return 1
  return n * fact(n - 1)

assert(fact(5) == 120)

def fact(n):
  acc = 1
  i = 1
  while i <= n:
    acc = acc * i
    i = i + 1
  return acc

assert(fact(5) == 120)

def fact(n):
  acc = 1
  i = 1
  def while_loop():
    nonlocal acc
    nonlocal i
    acc = acc * i
    i = i + 1
    if i <= n:
      while_loop()
  while_loop()
  return acc

assert(fact(5) == 120)

def fact(n):
  acc = 1
  i = 1
  def while_loop(acc, i):
    acc = acc * i
    i = i + 1
    if i <= n:
      return while_loop(acc, i)
    else:
      return acc
  return while_loop(acc, i)

assert(fact(5) == 120)
